src/Microsoft.DotNet.Interactive.HttpRequestParser/HttpRequestParser.cs

```diff
@@ -139,78 +147,102 @@ private void ConsumeCurrentTokenInto(HttpSyntaxNode node)
 
         private HttpRequestNode ParseRequest()
         {
+            var requestNode = new HttpRequestNode(
```

Comment:
We should have an early out here for the case where there are no more tokens (like we have in cases below)

Reply:
I suspect it's redundant with the methods this one calls but assuming there are bugs there, I'd prefer to find them by adding tests rather than adding code that has no coverage.

Reply:
I gave this a quick check and if we add a `MoreTokens` check and early out after each inner parse call, those branches are covered and all tests still pass. It looks to me like early outs will increase cyclomatic complexity and potentially negatively impact performance for the more cases when the code is well-formed.

Reply:
Yes, this is still one aspect of the pattern used in the Parse methods that I don't like as it would be easy to forget to add this early out when adding new constructs.

Not for the current PR, but it would be great to improve this and generally iron out the pattern that each Parse method follows so that it is easy to follow the same pattern in any Parse method for new constructs in the future.  As we discussed offline, we can explore making it impossible to return `null` by having `void` returning parse methods now that we allow nodes to be constructed before its children are parsed.

One question with that I think would be whether we will end up with cases where there is some leading / trailing trivia that we fail to round trip because we didn't have a node where it could be added. But I'm sure we can figure out some way to solve that...

Reply:
> Yes, this is still one aspect of the pattern used in the Parse methods that I don't like as it would be easy to forget to add this early out when adding new constructs.

I don't worry about this as long as the tests are solid, and I'm pretty confident the combinatorial tests will catch this.

Reply:
We would need to terminate the string with incomplete constructs to thoroughly test the early out where there are no more tokens (like the test we added yesterday where the string ended with a the `#` token (with missing comment body).

Many of these would also be in error scenarios which the combinatorial tests don't cover currently. But if we also extend the combinatorial testing to cover incorrect / incomplete syntax somehow that would catch this.

Reply:
I think we can use the combinatorial tests to find cases where there should be diagnostics but there aren't, at node-level granularity. I'll start on that soon.